### PR TITLE
refactor(protocol-designer): migrate i18n instances to use useTranslation hook

### PR DESCRIPTION
**Pull request by jerader**
closes RAUT-542

# Overview

This PR fixes all instances of i18n in the protocol-designer directory by using the `useTranslation` hook instead. This is because the previous way we were doing it wasn't properly implementing `react-i18next`. 

Additionally, several usages of the class components + connect function were refactored in order to use the `useTranslation` hook properly (this is part of a chore that needs to be done anyway to remove all the class components)

# Test Plan

Smoke test PD, try to go through every flow as much as possible and call out anything that doesn't work correctly or whitescreens.

Here is the sandbox to smoke test on: https://sandbox.designer.opentrons.com/pd_update-i18n/

# Changelog

- update i18ninit function
- update the `i18next` and `react-i18next` dependencies to be the latest
- update all instances of i18n in the components
- fix tests
- refactor several class components + connect functions

# Review requests

see test plan

# Risk assessment

medium-ish since it touches a bunch of code but it'll be internal

### CODE REVIEW

File: protocol-designer/package.json [hunk 1]
```diff
@@ -41,7 +41,7 @@
     "react-dnd": "6.0.0",
     "react-dnd-mouse-backend": "0.1.2",
     "react-dom": "18.2.0",
-    "react-i18next": "13.5.0",
+    "react-i18next": "14.0.0",
```
Comment (koji) [inline protocol-designer/package.json:44]: I guess If 13.5.0 works for PD, it will be good to keep the same version as well as `app`.
Also, we could move these two packages into the top level `package.json` like app-shell and app-shell-odd do.

**Comment by koji** [comment]
Seems that `minimum (5} uL`  part would have an issue.

2024-01-18 14_34_38-Opentrons Protocol Designer BETA

File: protocol-designer/src/components/BatchEditForm/NoBatchEditSharedSettings.tsx [hunk 2]
```diff
@@ -7,9 +7,10 @@ import {
   JUSTIFY_CENTER,
   SPACING_3,
 } from '@opentrons/components'
-import { i18n } from '../../localization'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/BatchEditForm/NoBatchEditSharedSettings.tsx]: you can move this line right after `react`.

File: protocol-designer/src/components/BatchEditForm/BatchEditMoveLiquid.tsx [hunk 3]
```diff
@@ -29,12 +28,14 @@ import { WellOrderOption } from '../../form-types'
 import formStyles from '../forms/forms.css'
 import styles from '../StepEditForm/StepEditForm.css'
 import buttonStyles from '../StepEditForm/ButtonRow/styles.css'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/BatchEditForm/BatchEditMoveLiquid.tsx]: you can move this line right after `react`.

File: protocol-designer/src/components/DeckSetup/LabwareOverlays/AdapterControls.tsx [hunk 4]
```diff
@@ -127,13 +126,11 @@ export const AdapterControlsComponents = (
...
             {!isOver && <Icon className={styles.overlay_icon} name="plus" />}
-            {i18n.t(
-              `deck.overlay.slot.${isOver ? 'place_here' : 'add_labware'}`
-            )}
+            {isOver ? 'Place Here' : 'Add Labware'}
           </a>
           <a className={styles.overlay_button} onClick={deleteLabware}>
             {!isOver && <Icon className={styles.overlay_icon} name="close" />}
-            {i18n.t('deck.overlay.edit.delete')}
+            {'Delete'}
```
Comment (koji) [inline protocol-designer/src/components/DeckSetup/LabwareOverlays/AdapterControls.tsx:133]: question
Any specific reason to pass the word directly?

File: protocol-designer/src/components/Hints/index.tsx [hunk 5]
```diff
@@ -1,84 +1,68 @@
...
   DeprecatedCheckboxField,
   Flex,
   OutlineButton,
   Text,
 } from '@opentrons/components'
-import { i18n } from '../../localization'
-import { actions as stepsActions } from '../../ui/steps'
-import { TerminalItemId } from '../../steplist'
+// import { actions as stepsActions } from '../../ui/steps'
+// import { TerminalItemId } from '../../steplist'
```
Comment (koji) [inline protocol-designer/src/components/Hints/index.tsx]: Need to keep these lines for the future development?

File: protocol-designer/src/components/StepEditForm/fields/MixFields.tsx [hunk 6]
```diff
@@ -1,15 +1,16 @@
 import * as React from 'react'
-import { i18n } from '../../../localization'
 import { CheckboxRowField, TextField } from './'
 import { FieldPropsByName } from '../types'
 import styles from '../StepEditForm.css'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/MixFields.tsx]: you can move this line right after react.

File: protocol-designer/src/components/StepEditForm/fields/PipetteField.tsx [hunk 7]
```diff
@@ -1,34 +1,25 @@
...
-import { connect } from 'react-redux'
-import { FormGroup, DropdownField, Options } from '@opentrons/components'
-import { i18n } from '../../../localization'
+import { FormGroup, DropdownField } from '@opentrons/components'
 import { selectors as stepFormSelectors } from '../../../step-forms'
-import { BaseState } from '../../../types'
 import styles from '../StepEditForm.css'
 import { FieldProps } from '../types'
+import { useTranslation } from 'react-i18next'
+import { useSelector } from 'react-redux'
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/PipetteField.tsx]: you can move these 2 lines right after `react`.

File: protocol-designer/src/components/StepEditForm/fields/TipPositionField/TipPositionModal.tsx [hunk 8]
```diff
@@ -45,13 +45,14 @@ const getErrorText = (args: {
   maxMmFromBottom: number
   minMmFromBottom: number
   isPristine: boolean
+  t: any
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/TipPositionField/TipPositionModal.tsx:48]: maybe adding a comment would be helpful to others?

File: protocol-designer/src/components/StepEditForm/fields/WellOrderField/WellOrderModal.tsx [hunk 9]
```diff
@@ -37,37 +37,53 @@ export interface WellOrderModalProps {
     firstValue?: WellOrderOption | null,
     secondValue?: WellOrderOption | null
   ) => void
+  t: any
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/WellOrderField/WellOrderModal.tsx:42]: oh, we need to upgrade packages...
https://www.i18next.com/overview/typescript

Could you leave a comment?

File: protocol-designer/src/components/StepEditForm/fields/WellSelectionField/WellSelectionField.tsx [hunk 10]
```diff
@@ -0,0 +1,112 @@
+import * as React from 'react'
+import { useDispatch, useSelector } from 'react-redux'
+import { useTranslation } from 'react-i18next'
+import { ALL, COLUMN } from '@opentrons/shared-data'
+import { selectors as stepFormSelectors } from '../../../../step-forms'
+import { FormGroup, InputField } from '@opentrons/components'
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/fields/WellSelectionField/WellSelectionField.tsx]:
```suggestion
import * as React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useTranslation } from 'react-i18next'
import { ALL, COLUMN } from '@opentrons/shared-data'
import { FormGroup, InputField } from '@opentrons/components'

import { selectors as stepFormSelectors } from '../../../../step-forms'
```

File: protocol-designer/src/components/StepEditForm/forms/PauseForm.tsx [hunk 11]
```diff
@@ -21,11 +20,13 @@ import { getSingleSelectDisabledTooltip } from '../utils'
 import styles from '../StepEditForm.css'
 
 import { StepFormProps } from '../types'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/forms/PauseForm.tsx]: you can move this line right after react.

File: protocol-designer/src/components/StepEditForm/forms/TemperatureForm.tsx [hunk 12]
```diff
@@ -2,13 +2,14 @@ import * as React from 'react'
 import { useSelector } from 'react-redux'
 import { selectors as uiModuleSelectors } from '../../../ui/modules'
 import { FormGroup } from '@opentrons/components'
-import { i18n } from '../../../localization'
 
 import { StepFormDropdown, RadioGroupField, TextField } from '../fields'
 import styles from '../StepEditForm.css'
 import { StepFormProps } from '../types'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/StepEditForm/forms/TemperatureForm.tsx]: same as the above

File: protocol-designer/src/components/alerts/ErrorContents.tsx [hunk 13]
```diff
@@ -1,9 +1,9 @@
 import * as React from 'react'
-import { i18n } from '../../localization'
 import { START_TERMINAL_ITEM_ID } from '../../steplist'
 import { KnowledgeBaseLink } from '../KnowledgeBaseLink'
 import { AlertLevel } from './types'
 import { TerminalItemLink } from '../steplist/TerminalItem'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/alerts/ErrorContents.tsx]: same as the above

File: protocol-designer/src/components/alerts/ErrorContents.tsx [hunk 14]
```diff
@@ -36,21 +37,20 @@ export const ErrorContents = (
       case 'NO_TIP_ON_PIPETTE':
         return (
           <>
-            {i18n.t(`alert.timeline.error.${props.errorType}.body1`)}
+            {t(`timeline.error.${props.errorType}.body1`)}
```
Comment (koji) [inline protocol-designer/src/components/alerts/ErrorContents.tsx:41]: maybe deconstruct `props` in the future?

File: protocol-designer/src/components/alerts/PDAlert.tsx [hunk 15]
```diff
@@ -1,10 +1,10 @@
 import * as React from 'react'
 import { AlertItem, OutlineButton } from '@opentrons/components'
-import { i18n } from '../../localization'
 // TODO: Ian 2019-03-27 the use of Component Library `Alert` is being
 // stretched beyond its intentions here, we should reconcile PD + Run App uses of Alert later
 import styles from './alerts.css'
 import { AlertData, AlertType } from './types'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/alerts/PDAlert.tsx]: same

File: protocol-designer/src/components/alerts/WarningContents.tsx [hunk 16]
```diff
@@ -1,8 +1,8 @@
 import * as React from 'react'
-import { i18n } from '../../localization'
 import { START_TERMINAL_ITEM_ID } from '../../steplist'
 import { AlertLevel } from './types'
 import { TerminalItemLink } from '../steplist/TerminalItem'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/alerts/WarningContents.tsx]: same

File: protocol-designer/src/components/labware/BrowseLabwareModal.tsx [hunk 17]
```diff
@@ -19,6 +18,7 @@ import { LabwareDefinition2 } from '@opentrons/shared-data'
 
 import modalStyles from '../modals/modal.css'
 import styles from './labware.css'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/labware/BrowseLabwareModal.tsx]: same

File: protocol-designer/src/components/DeckSetup/LabwareOverlays/AdapterControls.tsx (hunk 4 above)
Comment (jerader) [reply protocol-designer/src/components/DeckSetup/LabwareOverlays/AdapterControls.tsx:133]: i'll fix this in a follow up pr where i remove the usage of all connect functions. I think i was lazy here and just changed it to not using i18n 😓

File: protocol-designer/src/components/modals/CreateFileWizard/__tests__/CreateFileWizard.test.tsx [hunk 18]
```diff
@@ -1,4 +1,5 @@
 import * as React from 'react'
+import { i18n } from '../../../../localization'
```
Comment (koji) [inline protocol-designer/src/components/modals/CreateFileWizard/__tests__/CreateFileWizard.test.tsx]: can you move down this line like `app`' s tests do?

File: protocol-designer/src/components/modals/CreateFileWizard/__tests__/GoBack.test.tsx [hunk 19]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react'
-import i18n from 'i18next'
+import { i18n } from '../../../../localization'
```
Comment (koji) [inline protocol-designer/src/components/modals/CreateFileWizard/__tests__/GoBack.test.tsx]: can you move down this line like `app`' s tests do?

File: protocol-designer/src/components/modals/CreateFileWizard/__tests__/ModulesAndOtherTile.test.tsx [hunk 20]
```diff
@@ -1,5 +1,6 @@
 import * as React from 'react'
-import i18n from 'i18next'
+import { i18n } from '../../../../localization'
```
Comment (koji) [inline protocol-designer/src/components/modals/CreateFileWizard/__tests__/ModulesAndOtherTile.test.tsx]: can you move down this line like `app`' s tests do?

File: protocol-designer/src/components/StepEditForm/fields/WellOrderField/WellOrderModal.tsx (hunk 9 above)
Comment (jerader) [reply protocol-designer/src/components/StepEditForm/fields/WellOrderField/WellOrderModal.tsx:42]: good catch, for some reason the t was typing as any so i left it as any whenever i needed to use it as a prop. I'll link this url in the comment!

File: protocol-designer/src/components/modals/FilePipettesModal/PipetteFields.tsx [hunk 21]
```diff
@@ -28,8 +27,9 @@ import styles from './FilePipettesModal.css'
 import formStyles from '../../forms/forms.css'
 
 import type { PipetteName } from '@opentrons/shared-data'
-import { ThunkDispatch } from 'redux-thunk'
-import { BaseState } from '../../../types'
+import type { ThunkDispatch } from 'redux-thunk'
+import type { BaseState } from '../../../types'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/modals/FilePipettesModal/PipetteFields.tsx]: same

File: protocol-designer/src/components/modals/FilePipettesModal/index.tsx [hunk 22]
```diff
@@ -203,36 +325,33 @@ export class FilePipettesModal extends React.Component<Props, State> {
...
-                mount,
-                name: formPipette.pipetteName,
-                tiprackDefURI: formPipette.tiprackDefURI,
-              },
-            ]
-          : acc
-      },
-      []
-    )
+    // const newProtocolFields = values.fields
```
Comment (koji) [inline protocol-designer/src/components/modals/FilePipettesModal/index.tsx]: Need to keep these for another PR?

File: protocol-designer/src/components/modals/FileUploadMessageModal/FileUploadMessageModal.tsx [hunk 23]
```diff
@@ -1,10 +1,10 @@
-import { i18n } from '../../../localization'
 import * as React from 'react'
 import cx from 'classnames'
 import { AlertModal, OutlineButton } from '@opentrons/components'
 import modalStyles from '../modal.css'
 import { getModalContents } from './modalContents'
 import { FileUploadMessage } from '../../../load-file'
+import { useTranslation } from 'react-i18next'
```
Comment (koji) [inline protocol-designer/src/components/modals/FileUploadMessageModal/FileUploadMessageModal.tsx]: same

File: protocol-designer/src/initialize.ts [hunk 24]
```diff
@@ -1,12 +1,11 @@
-import { i18n } from './localization'
 import { selectors as loadFileSelectors } from './load-file'
 
 export const initialize = (store: Record<string, any>): void => {
   if (process.env.NODE_ENV === 'production') {
     window.onbeforeunload = (_e: unknown) => {
       // NOTE: the custom text will be ignored in modern browsers
       return loadFileSelectors.getHasUnsavedChanges(store.getState())
-        ? i18n.t('alert.window.confirm_leave')
+        ? 'Are you sure you want to leave? You will lose any unsaved changes.'
```
Comment (koji) [inline protocol-designer/src/initialize.ts:9]: don't need to add this text to a json file?

Comment (jerader) [reply protocol-designer/src/initialize.ts:9]: i am not sure how to import the useTranslation hook here since this const is used to initialize redux. I left a comment to go back to it though. After all the connect fns are refactored out, i can do an audit of the remainder text that is not in i18n yet.

File: protocol-designer/src/ui/modules/utils.ts [hunk 25]
```diff
@@ -61,6 +60,22 @@ export function getModuleUnderLabware(
     return location === moduleOnDeck.id
   })
 }
+
+export const getModuleShortNames = (type: ModuleType): string => {
+  switch (type) {
+    case 'heaterShakerModuleType':
```
Comment (koji) [inline protocol-designer/src/ui/modules/utils.ts:66]: maybe importing const from `shared-data` would be good?
```
import { 
  HEATERSHAKER_MODULE_TYPE,
  MAGNETIC_MODULE_TYPE,
  TEMPERATURE_MODULE_TYPE,
  THERMOCYCLER_MODULE_TYPE
} from ''
```

**Review by koji — APPROVED** [review]
Did the smoke test and didn't see any typical text display issue on PD.
Thank you for converting class component -> functional component! That is awesome.
